docs/how-to/how-to-run-docker-with-kata.md

```diff
@@ -0,0 +1,128 @@
+
+As `/var/lib/docker` is a `VOLUME` specified by DinD (i.e. the `docker` images tagged `*-dind`/`*-dind-rootless`), `docker` fill fail to start (or even worse, silently pick a worse storage driver like `vfs`) when started in a Kata Container. Special measures must be taken when running DinD-powered workloads in Kata Containers.
+
+## Workarounds/Solutions
+
+Thanks to various community contributions (see [issue references below](#references)) the following options, with various trade-offs have been uncovered:
+
+### Use a memory backed volume
+
+For small workloads (small container images, without much generated filesystem load), a memory-backed volume is sufficient. Kubernetes supports a variant of  [the `EmptyDir` volume][k8s-emptydir], which allows for memdisk-backed storage -- the [the `medium: Memory` ][k8s-memory-volume-type]. An example of a `Pod` using such a setup [was contributed](https://github.com/kata-containers/runtime/issues/1429#issuecomment-477385283), and is reproduced below:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dind
+spec:
+  runtimeClassName: kata
+  containers:
+  - name: dind
+    securityContext:
+      privileged: true
+    image: docker:18.06-dind
+    args: ["--storage-driver=overlay2"]
+    resources:
+      limits:
+        memory: "3G"
+    volumeMounts:
+      - mountPath: /var/run/
+        name: dockersock
+      - mountPath: /var/lib/docker
+        name: docker
+  volumes:
+    - name: dockersock
+      emptyDir: {}
+    - name: docker
+      emptyDir:
+        medium: Memory
+```
+
+Inside the container you can view the mount:
+
+```console
+/ # mount | grep lib\/docker
+tmpfs on /var/lib/docker type tmpfs (rw,relatime)
+```
+
+As is mentioned in the comment encapsulating this code, using volatile memory for container storage backing is a risky and could be possibly wasteful on machines that do not have a lot of RAM.
+
+### Use a `loop` mounted disk
+
+Using a `loop`-mounted disk that is provisioned shortly before starting of the container workload is another approach that yields good performance. It was also [contributed](https://github.com/kata-containers/runtime/issues/1888#issuecomment-739057384) and the portion related to mounting the loopback disk is reproduced below:
+
+```yaml
+spec:
+  containers:
+    - name: docker
+      image: docker:dind
+      command: ["sh", "-c"]
+      args:
+      - if [[ $(df -PT /var/lib/docker | awk 'NR==2 {print $2}') == virtiofs ]]; then
+          apk add e2fsprogs &&
+          truncate -s 20G /tmp/disk.img &&
+          mkfs.ext4 /tmp/disk.img &&
+          mount /tmp/disk.img /var/lib/docker; fi &&
+        dockerd-entrypoint.sh;
```

Comment:
It looks like `dockerd-entrypoint.sh` won't run at all unless virtiofs is being used. If that's correct, wouldn't it be clearer to error if virtiofs isn't being used?

Reply:
Yeah it would be -- this is an exact excerpt from the discussion in the comments that I thought made it pretty clear what was happening though it's not perfect. What I ended up using for example looks more like this:

```yaml
          command:
            - /bin/ash
            - -c
            - |
              # Check if the filesystem is formatted
              echo "[info] detecting existing filesystem for /var/lib/docker...";
              VAR_LIB_DOCKER_FILESYSTEM=$(df -PT /var/lib/docker | awk 'NR==2 {print $2}')
              echo -e "[info] detected existing filesystem [$EXISTING_FILESYSTEM]";

              if [ -z "$DOCKER_DISK_SIZE" ] ; then
                DOCKER_DISK_SIZE=20G
              fi
              echo -e "[info] docker loop-mounted disk size set to  [$DOCKER_DISK_SIZE]";

              DOCKER_DISK_PATH=/docker/disk.img

              if [[ "$VAR_LIB_DOCKER_FILESYSTEM" == "virtiofs" ]]; then
                if [[ ! -f "$DOCKER_DISK_PATH" ]]; then
                  echo -e "[info] setting up file system";
                  apk add e2fsprogs;
                  truncate -s $DOCKER_DISK_SIZE $DOCKER_DISK_PATH;
                  mkfs.ext4 $DOCKER_DISK_PATH;
                else
                  echo -e "[info] filesystem already present, no need to set up";
                fi

                echo -e "[info] mounting drive";
                mount $DOCKER_DISK_PATH /var/lib/docker;
              fi

              # Run original entrypoint
              echo -e "[info] starting docker entrypoint...";
              echo -e "";
              dockerd-entrypoint.sh --mtu=1300;
```

I don't think the snippet has to be perfect but in this case what I have there was more to quote the important bit from the mentioned contribution.